File: src/gestorAplicacion/usuario/Profesor.java
/*
 * Autores:
 *  EFRAÍN GÓMEZ RAMÍREZ
 *  LIBARDO JOSÉ NAVARRO PEDROZO
 *  ANA SOFÍA GÓMEZ ZAPATA
 *  SEBASTIÁN OCAMPO GALVIS
 */

package gestorAplicacion.usuario;
import java.util.ArrayList;
import gestorAplicacion.administracion.*;
import java.io.Serializable;

public class Profesor implements Serializable{
    private String nombre;
    private String facultad;
    private Horario horario;
    private ArrayList<Materia> materiasDadas = new ArrayList<Materia>(10);
    private ArrayList<Grupo> grupos = new ArrayList<Grupo>();
    private static ArrayList<Profesor> profesores =new ArrayList<Profesor>();
    private static final long serialVersionUID = 1L;
    
    public Profesor(String nombre, String facultad, Horario horario, ArrayList<Materia> materiasDadas, ArrayList<Grupo> grupos){
        this.nombre = nombre;
        this.facultad = facultad;
        this.horario = horario;
        this.materiasDadas = materiasDadas;
        this.grupos = grupos;
        Profesor.profesores.add(this);
    }

    public Profesor(String nombre, String facultad, Horario horario, ArrayList<Materia> materiasDadas){
        this.nombre = nombre;
        this.facultad = facultad;
        this.horario = horario;
        this.materiasDadas = materiasDadas;
        Profesor.profesores.add(this);
    }
    
    public Profesor(String nombre, String facultad, ArrayList<Materia> materiasDadas){
        this.nombre = nombre;
        this.facultad = facultad;
        this.horario = new Horario();
        this.materiasDadas = materiasDadas;
        Profesor.profesores.add(this);
    }
    
    public void vincularGrupo(Grupo g) {
    	this.grupos.add(g);
    	this.horario.ocuparHorario(g.getHorario(), g);
    }
    
    public void desvincularGrupo(Grupo g){
        if (grupos.contains(g)){
            int indice = grupos.indexOf(g);
            ArrayList<String> horLibre = grupos.get(indice).getHorario();
            this.horario.liberarHorario(horLibre);
            grupos.remove(indice);
        }
    }

    public boolean daMateria(String nombre) {
    	for(Materia materia:this.getMateriasDadas()) {
    		if(materia.getNombre().equals(nombre)) {
    			return true;
    		}
    	}
    	return false;
    }

    public static boolean recomendarEstudiante(Estudiante estudiante){
        boolean bool = false;
        for (Profesor profesor : Profesor.getProfesores()){
            int chance = 0;
            int suerte = (int)(Math.random()*10+1);

            for(Grupo grupo : estudiante.getGruposVistos()){
                if (grupo.getProfesor().getNombre().equals(profesor.getNombre()) == true){
                    chance += 5;
                    break;
                }
            }
            
            if (estudiante.getFacultad().equals(profesor.getFacultad()) == true){
                chance += 3;
            }
            if (chance >= suerte){
                bool= true;
            }  
        }
        return bool;
    }
    
    public static String mostrarProfesores() {
    	String r = "";
    	int i = 1;
    	for(Profesor profesor:Profesor.profesores) {
    		r += (i++)+". "+profesor.getNombre()+". Materias: ";
    		for (Materia materia:profesor.getMateriasDadas()) {
    			if (profesor.getMateriasDadas().indexOf(materia) == profesor.getMateriasDadas().size()-1) {
    				r += materia.getNombre()+".\n";
    			}
    			else {
    				r += materia.getNombre()+", ";
    			}
    		}
    	}
    	return r;
    }
    
    public static ArrayList<Profesor> profesoresDeMateria(String nombre) {
    	ArrayList<Profesor> profes = new ArrayList<Profesor>();
    	for (Profesor profesor:Profesor.getProfesores()) {
    		if(profesor.daMateria(nombre)&&!profes.contains(profesor)) {
    			profes.add(profesor);
    		}
    	}
    	return profes;
    }
    
    public static String mostrarProfesMateria(String nombre) {
    	String r = "";
    	int i = 1;
    	ArrayList<Profesor> profes = profesoresDeMateria(nombre);
    	for (Profesor profesor:profes) {
    		r += (i++)+". "+profesor.getNombre()+".\n"; 
    	}
    	return r;
    }

    public String getNombre() {
        return nombre;
    }
    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getFacultad() {
        return facultad;
    }
    public void setFacultad(String facultad) {
        this.facultad = facultad;
    }

    public ArrayList<Materia> getMateriasDadas() {
        return materiasDadas;
    }
    public void setMateriasDadas(ArrayList<Materia> materiasDadas) {
        this.materiasDadas = materiasDadas;
    }

    public ArrayList<Grupo> getGrupos() {
        return grupos;
    }
    public void setGrupos(ArrayList<Grupo> grupos) {
        this.grupos = grupos;
    }

    public static ArrayList<Profesor> getProfesores() {
        return profesores;
    }
    public static void setProfesores(ArrayList<Profesor> profesores) {
        Profesor.profesores = profesores;
    }
    
    public void setHorario(Horario horario) {
    	this.horario = horario;
    }
    
    public Horario getHorario() {
    	return this.horario;
    }

}

File: Python/src/gestorAplicacion/usuario/Profesor.py
# Autores:
# EFRAÍN GÓMEZ RAMÍREZ
# LIBARDO JOSÉ NAVARRO PEDROZO
# ANA SOFÍA GÓMEZ ZAPATA
# SEBASTIÁN OCAMPO GALVIS
from gestorAplicacion.administracion.Horario import Horario
import random


class Profesor:
    _profesores = []

    def __init__(self, nombre, facultad, materiasDadas, horario=Horario(), grupos=[]):
        self._nombre = nombre
        self._facultad = facultad
        self._materiasDadas = materiasDadas
        self._horario = horario
        self._grupos = grupos if grupos is not None else []
        Profesor._profesores.append(self)

    def vincularGrupo(self, g):
        self._grupos.append(g)
        self._horario.ocuparHorario(g, g.getHorario())

    def desvincularGrupo(self, g):
        if g in self._grupos:
            indice = self._grupos.index(g)
            horaLibre = self._grupos[indice].getHorario()
            self._horario.liberarHorario(horaLibre)
            self._grupos.remove(g)

    def daMateria(self, nombre):
        for materia in self.getMateriasDadas():
            if materia.getNombre() == nombre:
                return True
        return False

    @classmethod
    def recomendarEstudiante(cls, estudiante):
        for profesor in cls._profesores:
            chance = 0
            suerte = random.randint(1, 10)
            for grupo in estudiante.getGruposVistos():
                if grupo.getProfesor().getNombre() == profesor.getNombre():
                    chance += 5
                    break
            if estudiante.getFacultad() == profesor.getFacultad():
                chance += 3
            if chance >= suerte:
                return True
        return False

    @classmethod
    def mostrarProfesores(cls):
        r = ""
        i = 1
        for profesor in cls._profesores:
            r += f"{i}. {profesor.getNombre()}. Materias: "
            for materia in profesor.getMateriasDadas():
                if (
                    profesor.getMateriasDadas().index(materia)
                    == len(profesor.getMateriasDadas()) - 1
                ):
                    r += f"{materia.getNombre()}.\n"
                else:
                    r += f"{materia.getNombre()}, "
            i += 1
        return r

    @classmethod
    def profesoresDeMateria(cls, nombre):
        profes = []
        for profesor in cls._profesores:
            if profesor.daMateria(nombre) and profesor not in profes:
                profes.append(profesor)
        return profes
    
    @classmethod
    def nombresProfesDeMateria(cls, materia):
        profes = []
        for profesor in cls._profesores:
            if profesor.daMateria(materia) and profesor not in profes:
                profes.append(profesor.getNombre())
        return profes

    @classmethod
    def encontrarProfe(cls, profesor):
        for profe in Profesor._profesores:
            if profe.getNombre()== profesor:
                return profe

    @classmethod
    def mostrarProfesMateria(cls, nombre):
        r = ""
        i = 1
        profes = cls._profesoresDeMateria(nombre)
        for profesor in profes:
            r += f"{i}. {profesor.getNombre()}.\n"
            i += 1
        return r

    def getNombre(self):
        return self._nombre

    def setNombre(self, nombre):
        self._nombre = nombre

    def getFacultad(self):
        return self._facultad

    def setFacultad(self, facultad):
        self._facultad = facultad

    def getMateriasDadas(self):
        return self._materiasDadas

    def setMateriasDadas(self, materiasDadas):
        self._materiasDadas = materiasDadas

    def getGrupos(self):
        return self._grupos

    def setGrupos(self, grupos):
        self._grupos = grupos

    def setHorario(self, horario):
        self._horario = horario

    def getHorario(self):
        return self._horario

    @classmethod
    def getProfesores(cls):
        return cls._profesores

    @classmethod
    def setProfesores(cls, profesores):
        cls._profesores = profesores
